Feat/contribution tree

will7200 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and
review the checklist below.

Contributors guide: ./CONTRIBUTING.md
-->

## Checklist
<!--
Remove items that do not apply. For completed items, change [ ] to [x] or you can click the checkboxes once your 
pull-request is published.
-->

- [x] Create new `NavigatorWidget` class for contribution tree, and make it available in this row:

- [x] Design decision on `tags` system. Tags are a part of [our fundamental data schema](https://github.com/brightway-lca/bw_interface_schemas/blob/5fb1d40587aec2a4bb2248505550fc883a91c355/bw_interface_schemas/lci.py#L68), and are designed to be categorical (quantitative values would normally be a `property`).
- [x] Add ability to add tags to processes, either in the existing edit interface or in a new pop-up. This interface should autocomplete against existing tag systems already in the data; we might want to cache these on the `Database` object instead of scanning the whole database when doing any initial load.
- [x] Implemented aggregation algorithm by `tags` system to the results. Handle missing `tags` elegantly.
- [x] Write initial graph traversal logic in new NavigatorWidget` class 
- [x] Write click event handling logic in `NavigatorWidget` class 
- [ ] User acceptance and change requests


<!--
NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

will7200 committed:
fix: html buttons on macos

will7200 committed:
fix: html buttons on macos

will7200 committed:
fix: on click for graph navigator

will7200 committed:
fix: on click for graph navigator

will7200 committed:
feat: add tree navigator

will7200 committed:
feat: add tree navigator

cmutel commented on activity_browser/application.py:
```diff
@@ -41,14 +41,14 @@ def deleteLater(self):
-if QSysInfo.productType() in ["arch", "nixos"]:
-    os.environ["QTWEBENGINE_CHROMIUM_FLAGS"] = "--no-sandbox"
+if QSysInfo.productType() in ["arch", "nixos", "osx"]:
+    os.environ["QTWEBENGINE_CHROMIUM_FLAGS"] = "{} --no-sandbox".format(os.getenv("QTWEBENGINE_CHROMIUM_FLAGS"))
```
This needs a bit of explanation on why and how to set the envvar.

will7200 replied:
added the why

cmutel commented on activity_browser/ui/web/tree_navigator.py:
```diff
@@ -0,0 +1,602 @@
+        }
+
+    @classmethod
+    def calculate(
```
I'd be a lot happier if we could incorporate missing functionality into `bw_graph_tools`; the AB has done a lot of stuff on it's own in the past, and we now have to fix the messes they created.

will7200 replied:
moved to bw_graph_tools and cleaned it up a bit as well
https://github.com/brightway-lca/bw_graph_tools/pull/20

cmutel commented:
Thanks @will7200 .Can we have a call to talk about this? Shouldn't take too long. Maybe in the next 6 hours?

will7200 committed:
feat: update graph elements

1. update dagre to v6
2. add resize watcher for graph window
3. add zoom to element functionality
4. fix dynamic expanding of nodes indicator not working
5. update toggle of node to recursively add nodes if available

will7200 committed:
feat: update graph elements

1. update dagre to v6
2. add resize watcher for graph window
3. add zoom to element functionality
4. fix dynamic expanding of nodes indicator not working
5. update toggle of node to recursively add nodes if available

will7200 committed:
feat: add node zoom to selection

fix dynamic indicator for different orientations

will7200 committed:
feat: add node zoom to selection

fix dynamic indicator for different orientations

will7200 committed:
refactor: use bw_graph_tools graph traversal

will7200 committed:
refactor: use bw_graph_tools graph traversal

will7200 committed:
feat: add tag editor for processes

will7200 committed:
feat: add tag editor for processes

will7200 committed:
refactor: add node go to other navigators

will7200 committed:
refactor: add node go to other navigators

will7200 committed:
feat: add auto complete for tag names

will7200 committed:
feat: add auto complete for tag names

will7200 commented:
@cmutel 
For the tag types, since they are aimed at being categorical data types, what data types do we want to support?
The tags support json values so that consist of numbers, strings, arrays, and objects.
Currently I implemented support for: 
+ `int`
+ `float`
+ `string`
+ `date (string underneath)`

So we can drop the number ones (int and float). Do we need to support others ones? How much benefit would supporting lists and objects?

cmutel commented:
> So we can drop the number ones (int and float). Do we need to support others ones? How much benefit would supporting lists and objects?

A good question - probably the schema is too loose here. I can see a good argument for `int`, in some numbered classification scheme, but am struggling to see how `float` would make sense (especially considering underlying floating point mechanics). I also agree that datetimes smell a bit fishy, but if converted to ISO 8601 then that seems OK.

**Definitely not** objects including lists or any other nonsense. Special cases aren't special enough.

will7200 committed:
feat: add insert and remove rows functionality for pandas model

will7200 committed:
feat: add insert and remove rows functionality for pandas model

will7200 committed:
chore: update to latest graph api

will7200 committed:
chore: update to latest graph api

will7200 committed:
feat: add tag ui element for tree

will7200 committed:
feat: add tag ui element for tree

will7200 committed:
feat: add graph grouping by tags

will7200 committed:
feat: add graph grouping by tags

will7200 committed:
chore: api updates

will7200 committed:
test: add for tag editor

will7200 commented:
@cmutel Ready for review
will require https://github.com/brightway-lca/bw_graph_tools/pull/20 to be merged for tests to pass

cmutel committed:
Factorize for better speed on umfpack

cmutel committed:
Bump dependencies

cmutel committed:
Convert inputs to ints

cmutel committed:
Merge branch 'feat/contribution-tree' of https://github.com/cauldron/activity-browser into feat/contribution-tree

cmutel committed:
Factorize in tree expansion LCA

cmutel committed:
Fix for bw_graph_tools 0.5

cmutel committed:
Cutoff values must be ints

cmutel committed:
3.0.dev8 release
